kernel/src/hil/spi.rs

```diff
@@ -40,6 +40,72 @@ pub enum ClockPhase {
     SampleTrailing,
 }
 
+pub mod util {
+
+    #[derive(Copy, Clone)]
+    pub enum Polarity {
+        High,
+        Low,
+    }
+
+    mod private {
+        pub trait Sealed {}
+    }
+
+    pub trait ChipSelect: private::Sealed {
+        const POLARITY: Polarity;
+    }
+
+    pub enum ActiveLow {}
+    pub enum ActiveHigh {}
+
+    impl private::Sealed for ActiveLow {}
+    impl private::Sealed for ActiveHigh {}
+
+    impl ChipSelect for ActiveLow {
+        const POLARITY: Polarity = Polarity::Low;
+    }
+
+    impl ChipSelect for ActiveHigh {
+        const POLARITY: Polarity = Polarity::High;
+    }
+
+    pub trait IntoChipSelect<T, ChipSelectActivePolarity> {
+        fn into_cs(self) -> T;
+    }
+
+    #[derive(Copy, Clone)]
+    pub struct ChipSelectPolar<P> {
+        pub pin: P,
+        pub polarity: Polarity,
+    }
+
+    impl<P, A: ChipSelect> IntoChipSelect<ChipSelectPolar<P>, A> for P {
+        fn into_cs(self) -> ChipSelectPolar<P> {
+            ChipSelectPolar {
+                pin: self,
+                polarity: A::POLARITY,
+            }
+        }
+    }
+
+    impl<P: crate::hil::gpio::Output> ChipSelectPolar<P> {
+        pub fn deactivate(&self) {
+            match self.polarity {
+                Polarity::Low => self.pin.set(),
+                Polarity::High => self.pin.clear(),
+            }
+        }
+
+        pub fn activate(&self) {
+            match self.polarity {
+                Polarity::Low => self.pin.clear(),
+                Polarity::High => self.pin.set(),
+            }
+        }
+    }
+}
```

Comment:
This would have to be with the assumption that those controllers _all_ implement active low _only_ (likely?). See the revised version of SAM4L, which I think is better anyway and requires the implementation to declare explicitly what polarity it supports.